### PR TITLE
SDIT-898 Improve POST /api/bookings/offence-history

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -53,7 +53,7 @@ dependencies {
   implementation("commons-io:commons-io:2.13.0")
   implementation("org.apache.commons:commons-text:1.10.0")
   implementation("com.oracle.database.jdbc:ojdbc10:19.19.0.0")
-  implementation("org.hibernate.orm:hibernate-community-dialects:6.2.2.Final")
+  implementation("org.hibernate.orm:hibernate-community-dialects")
```
Comment (steverendell) [inline build.gradle.kts:56]: Suggestion from earlier which keeps version in step with spring

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingResource.java [hunk 2]
```diff
@@ -656,7 +656,8 @@ public List<OffenceHistoryDetail> getOffenceHistory(@PathVariable("offenderNo")
     @Operation(summary = "Offence histories.", description = "Offence histories for a set of booking ids.")
     @PostMapping("/offence-history")
     @PreAuthorize("hasAnyRole('SYSTEM_USER','VIEW_PRISONER_DATA')")
-    public List<OffenceHistoryDetail> getOffenceHistoryForBookings(@RequestBody @Parameter(description = "The booking ids", required = true) final Set<Long> bookingIds){
+    @SlowReportQuery
+    public List<OffenceHistoryDetail> getOffenceHistoryForBookings(@RequestBody @Parameter(description = "The booking ids", required = true) final Set<Long> bookingIds) {
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingResource.java:660]: can run on replica, slow, used by batches and readonly

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java [hunk 3]
```diff
@@ -115,10 +144,6 @@ public boolean isActive() {
         return ACTIVE.equals(chargeStatus);
     }
 
-    public void setMostSeriousFlag(String mostSeriousFlag) {
-        this.mostSeriousFlag = mostSeriousFlag;
-    }
-
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java:121]: not used

File: src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java [hunk 4]
```diff
@@ -390,7 +390,7 @@ public Page<VisitWithVisitors> getBookingVisitsWithVisitor(final VisitInformatio
                         .flatMap(id -> allContacts.getOrDefault(id, Optional.empty()))
                         .map(OffenderContactPerson::getRelationshipType);
 
-                    final var visitorsList = getVisitors(allVisitors, allContacts, visitInformation.getVisitId(), filter.getBookingId());
+                    final var visitorsList = getVisitors(allVisitors, allContacts, visitInformation.getVisitId());
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java:393]: unused parameter (intellij warning)

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java [hunk 5]
```diff
@@ -35,6 +38,32 @@
 @Table(name = "OFFENDER_CHARGES")
 @ToString(exclude = {"offenderBooking", "offenderCourtCase"})
 @With
+@NamedEntityGraph(
+    name = "charges-details",
+    attributeNodes = {
+        @NamedAttributeNode(value = "offenderCourtCase", subgraph = "court-case"),
+        @NamedAttributeNode(value = "offence"),
+        @NamedAttributeNode(value = "resultCodeOne"),
+        @NamedAttributeNode(value = "resultCodeTwo"),
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java:47]: I didn't realise this needs doing for ManyToOnes as well. Surprised Hibernate doesn't do a join on them.

**Review by mikehalmamoj — APPROVED** [review]

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java (hunk 5 above)
Comment (steverendell) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderCharge.java:47]: Indeed, actually some of these joins were done by hibernate 6.1, but not 6.2

**Pull request merged**